File: 0714. Best Time to Buy and Sell Stock with Transaction Fee/Solution0714.cpp
#include<vector>
#include<algorithm>
#include<iostream>

using namespace std;

class Solution0714 {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int buy = -prices[0];
        int sell = 0;
        for (int i = 1; i < prices.size(); i++) {
            buy = max(buy, sell - prices[i]);
            sell = max(sell, buy + prices[i] - fee);
        }
        return sell;
    }
};

File: 0714. Best Time to Buy and Sell Stock with Transaction Fee/Solution0714.py
from typing import List


class Solution0714:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        buy, sell = -prices[0], 0
        for i in range(1, len(prices)):
            buy = max(buy, sell - prices[i])
            sell = max(sell, buy + prices[i] - fee)
        return sell